### PR TITLE
Project items

### PR DESCRIPTION
**Pull request by MichaelLiddiard**
Reworked project system so the majority of physical file code is moved to the clients.

### CODE REVIEW

File: src/Projects/Model/ItemModel.cs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        [JsonProperty]
+        public Guid? ParentItemId { get; private set; }
+        [JsonProperty]
+        public string Type { get; private set; } = "folder";
+
+        [JsonIgnore]
+        public List<ItemModel> Children { get; set; } = new List<ItemModel>();
+
+        [JsonIgnore]
+        public bool LocalOnly { get; set; } = false;
```
Comment (jon-r-abbott) [inline src/Projects/Model/ItemModel.cs:26]: Should the set be private?

File: src/Projects/Model/ItemModel.cs [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        {
+            if (parent == null)
+                throw new ArgumentNullException();
+
+            Name = name;
+            PhysicalName = physicalname;
+            ProjectId = parent.Id;
+            Type = type;
+            LocalOnly = true;            
+        }
```
Comment (jon-r-abbott) [inline src/Projects/Model/ItemModel.cs]: These constructors don't smell right. As discussed, no action as work in progress.

File: src/Projects/Model/ItemModel.cs [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+        [JsonIgnore]
+        public ItemModel Parent { get; set; }
+
+        [JsonConstructor]
+        private ItemModel()
+        { }
+
+        internal ItemModel(string name, string physicalname, ItemModel parent, string type, bool localOnly)
+        {
```
Comment (jon-r-abbott) [inline src/Projects/Model/ItemModel.cs:40]: Any reason why this constructor isn't check for null parent?

File: src/Projects/ProjectService.cs [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+        public async Task<IEnumerable<ProjectModel>> GetAllProjects()
+        {
+            if (_projectModels == null)
+            {
+                _projectModels = new ObservableCollection<ProjectModel>();
+            }
+
+            try
+            {
+                var task = await _auth.GetAuthenticatedClient().GetAsync(PROJECT_ENDPOINT);
```
Comment (jon-r-abbott) [inline src/Projects/ProjectService.cs:44]: Not checking HttpResponseMessage for IsSuccessStatusCode. No action at the moment as continue will throw exception.

File: src/Projects/ProjectService.cs [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+                var jsonString = await task.Content.ReadAsStringAsync();
+                var result = JsonConvert.DeserializeObject<IEnumerable<ProjectModel>>(jsonString);
+                _projectModels = new ObservableCollection<ProjectModel>(result);
+                foreach (ProjectModel pm in _projectModels)
+                {
+                    pm.PropertyChanged += ProjectModelChanged;
+                }
```
Comment (jon-r-abbott) [inline src/Projects/ProjectService.cs:51]: As discussed, doesn't feel right as GetProject doesn't hook up the events. No action as wip.

File: src/Projects/ProjectService.cs [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+            }
+
+            return _projectModels;
+        }
+
+        public async Task<ProjectModel> GetProject(Guid id)
+        {
+            try
+            {
+                var task = await _auth.GetAuthenticatedClient().GetAsync($"{PROJECT_ENDPOINT}/{id}");
```
Comment (jon-r-abbott) [inline src/Projects/ProjectService.cs:67]: Not checking HttpResponseMessage for IsSuccessStatusCode. No action at the moment as continue will throw exception.

File: src/Projects/ProjectService.cs [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+        }
+
+        public async Task<ItemModel> CreateItem(string name, string physicalname, ItemModel parent, string type)
+        {
+            ItemModel im = new ItemModel(name, physicalname, parent, type);
+            
+            var stringPayload = await Task.Run(() => JsonConvert.SerializeObject(im));
+            var httpContent = new StringContent(stringPayload, Encoding.UTF8, "application/json");
+
+            var task = await _auth.GetAuthenticatedClient().PostAsync($"{ITEM_ENDPOINT}", httpContent);
```
Comment (jon-r-abbott) [inline src/Projects/ProjectService.cs:93]: Not checking HttpResponseMessage for IsSuccessStatusCode. Needs action, as will return even if not successful ItemModel.

File: src/Projects/ProjectService.cs [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+            }
+            else
+            {
+                return projectModel;
+            }
+        }
+
+        public async Task DeleteItem(Guid id)
+        {
+            var task = await _auth.GetAuthenticatedClient().DeleteAsync($"{ITEM_ENDPOINT}/{id}");
```
Comment (jon-r-abbott) [inline src/Projects/ProjectService.cs:135]: task variable unused - remove or discard "_"

Not checking HttpResponseMessage for IsSuccessStatusCode. No action needed at the moment.

File: src/Projects/ProjectService.cs [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+            var task = await _auth.GetAuthenticatedClient().PostAsync($"{ITEM_ENDPOINT}", httpContent);
+            return im;
+        }
+        
+        private async void ProjectModelChanged(object sender, PropertyChangedEventArgs e)
+        {
+            ProjectModel pm = sender as ProjectModel;
+            if(e.PropertyName.Equals(nameof(ProjectModel.Favourite)))
+            {
+                var task = await _auth.GetAuthenticatedClient().PutAsync(PROJECT_ENDPOINT + $"/{pm.Id}/Favourite", null);
```
Comment (jon-r-abbott) [inline src/Projects/ProjectService.cs:103]: task variable unused - remove or discard "_"

Not checking HttpResponseMessage for IsSuccessStatusCode. No action needed at the moment.

File: src/Projects/ProjectService.cs [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+                return result;
+            }
+            catch (Exception e)
+            {
+                Console.WriteLine(e);
+                throw;
+            }
+        }
+
+        public async Task<ItemModel> CreateItem(string name, string physicalname, ItemModel parent, string type)
```
Comment (jon-r-abbott) [inline src/Projects/ProjectService.cs]: physicalname - parameter name lowerCamelCase

File: src/Projects/ProjectService.cs [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+            var stringPayload = await Task.Run(() => JsonConvert.SerializeObject(im));
+            var httpContent = new StringContent(stringPayload, Encoding.UTF8, "application/json");
+
+            var task = await _auth.GetAuthenticatedClient().PostAsync($"{ITEM_ENDPOINT}", httpContent);
+            return im;
+        }
+        
+        private async void ProjectModelChanged(object sender, PropertyChangedEventArgs e)
+        {
+            ProjectModel pm = sender as ProjectModel;
```
Comment (jon-r-abbott) [inline src/Projects/ProjectService.cs]: Possible null reference

File: src/Projects/Model/ProjectModel.cs [hunk 12]
```diff
@@ -0,0 +1,158 @@
+﻿using Newtonsoft.Json;
+using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Linq;
+using System.Text;
+using DocumentFormat.OpenXml.Spreadsheet;
```
Comment (jon-r-abbott) [inline src/Projects/Model/ProjectModel.cs]: Clear up unused using directives

File: src/IMessageProvider.cs [hunk 13]
```diff
@@ -26,5 +26,9 @@ public interface IMessageProvider
         /// <param name="message">Message to be displayed</param>
         /// <returns>Awaitable task</returns>
         Task ShowError(string message);
+
+        Task ShowMessage(string message, string title);
```
Comment (jon-r-abbott) [inline src/IMessageProvider.cs:36]: Missing XML comment
Unused - assume this is for later use?

File: src/IMessageProvider.cs [hunk 14]
```diff
@@ -26,5 +26,9 @@ public interface IMessageProvider
...
+
+        Task<bool> ShowConfirmDialog(string message, string title = "Confirm");
```
Comment (jon-r-abbott) [inline src/IMessageProvider.cs:44]: Missing XML comment

File: src/Projects/IProjectService.cs [hunk 15]
```diff
@@ -0,0 +1,19 @@
...
+{
+    public interface IProjectService
+    {
+        Task<IEnumerable<ProjectModel>> GetAllProjects();
+
+        Task<ProjectModel> GetProject(Guid id);
+
+        Task<ItemModel> CreateItem(string name, string physicalname, ItemModel parent, string type);
+
+        Task DeleteItem(Guid id);
```
Comment (jon-r-abbott) [inline src/Projects/IProjectService.cs]: Missing XML comments

File: src/Projects/IProjectService.cs [hunk 16]
```diff
@@ -0,0 +1,19 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
```
Comment (jon-r-abbott) [inline src/Projects/IProjectService.cs]: Unnecessary using directive

File: src/Projects/IProjectService.cs [hunk 17]
```diff
@@ -0,0 +1,19 @@
...
+
+namespace Jpp.Common.Backend.Projects
+{
+    public interface IProjectService
+    {
+        Task<IEnumerable<ProjectModel>> GetAllProjects();
+
+        Task<ProjectModel> GetProject(Guid id);
+
+        Task<ItemModel> CreateItem(string name, string physicalname, ItemModel parent, string type);
```
Comment (jon-r-abbott) [inline src/Projects/IProjectService.cs:31]: physicalname - parameter name lowerCamelCase

File: src/Projects/Model/IPhysicalFactory.cs [hunk 18]
```diff
@@ -0,0 +1,11 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
```
Comment (jon-r-abbott) [inline src/Projects/Model/IPhysicalFactory.cs]: Unnecessary using directive

File: src/Projects/Model/IPhysicalFactory.cs [hunk 19]
```diff
@@ -0,0 +1,11 @@
...
+
+namespace Jpp.Common.Backend.Projects.Model
+{
+    public interface IPhysicalFactory
+    {
+        IPhysicalProject CreateProject(ProjectModel projectModel, IProjectService _projects);
```
Comment (jon-r-abbott) [inline src/Projects/Model/IPhysicalFactory.cs]: Missing XML comments
_projects - parameter name lowerCamelCase

File: src/Projects/Model/IPhysicalItem.cs [hunk 20]
```diff
@@ -0,0 +1,18 @@
...
+{
+    public interface IPhysicalItem
+    {
+        bool Unexpected { get;  }
+        
+        bool Exists { get; set; }
+
+        Task Create(IStorageProvider storage);
+
+        ItemModel LogicalItem { get; set; }
```
Comment (jon-r-abbott) [inline src/Projects/Model/IPhysicalItem.cs:13]: Missing XML comments

File: src/Projects/Model/IPhysicalItem.cs [hunk 21]
```diff
@@ -0,0 +1,18 @@
+﻿using System;
+using System.Collections.Generic;
```
Comment (jon-r-abbott) [inline src/Projects/Model/IPhysicalItem.cs]: Unnecessary using directives

File: src/Projects/Model/IPhysicalProject.cs [hunk 22]
```diff
@@ -0,0 +1,21 @@
+﻿using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Text;
+using System.Threading.Tasks;
```
Comment (jon-r-abbott) [inline src/Projects/Model/IPhysicalProject.cs]: Unnecessary using directives

File: src/Projects/Model/IPhysicalProject.cs [hunk 23]
```diff
@@ -0,0 +1,21 @@
...
+    {
+        bool FolderExists { get; }
+        string RootPath { get; set; }
+        ProjectModel LogicalProject { get; set; }
+
+        Task BuildItemTree();
+        Task CreateFolders();
+        Task OpenFolders();
+
+        IPhysicalItem ConvertLogicalItemToPhysical(ItemModel item);
```
Comment (jon-r-abbott) [inline src/Projects/Model/IPhysicalProject.cs:16]: XML comments?

File: src/Projects/Model/ItemModel.cs [hunk 24]
```diff
@@ -0,0 +1,75 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using DocumentFormat.OpenXml.Office2013.PowerPoint;
```
Comment (jon-r-abbott) [inline src/Projects/Model/ItemModel.cs]: Unnecessary using directives

File: src/Projects/Model/ItemModel.cs [hunk 25]
```diff
@@ -0,0 +1,75 @@
...
+        public IPhysicalItem PhysicalItem { get; set; }
+
+        [JsonIgnore]
+        public ItemModel Parent { get; set; }
+
+        [JsonConstructor]
+        private ItemModel()
+        { }
+
+        internal ItemModel(string name, string physicalname, ItemModel parent, string type, bool localOnly)
```
Comment (jon-r-abbott) [inline src/Projects/Model/ItemModel.cs]: physicalname - parameter name lowerCamelCase

File: src/Projects/Model/ItemModel.cs [hunk 26]
```diff
@@ -0,0 +1,75 @@
...
+        {
+            Name = name;
+            PhysicalName = physicalname;
+            ProjectId = parent.ProjectId;
+            ParentItemId = parent.Id;
+            Type = type;
+            LocalOnly = localOnly;
+        }
+
+        public ItemModel(string name, string physicalname, ItemModel parent, string type)
```
Comment (jon-r-abbott) [inline src/Projects/Model/ItemModel.cs]: physicalname - parameter name lowerCamelCase

File: src/Projects/Model/ItemModel.cs [hunk 27]
```diff
@@ -0,0 +1,75 @@
...
+
+            Name = name;
+            PhysicalName = physicalname;
+            ProjectId = parent.ProjectId;
+            ParentItemId = parent.Id;
+            Type = type;
+            LocalOnly = true;
+        }
+
+        public ItemModel(string name, string physicalname, ProjectModel parent, string type)
```
Comment (jon-r-abbott) [inline src/Projects/Model/ItemModel.cs]: physicalname - parameter name lowerCamelCase

File: src/Projects/Model/ProjectModel.cs [hunk 28]
```diff
@@ -0,0 +1,158 @@
...
+            container.Add(itemModel);
+            itemModel.PhysicalItem = PhysicalProject?.ConvertLogicalItemToPhysical(itemModel);
+            BuildChildTree(itemModel);
+        }
+
+        public ItemModel GetItemByPath(string fullpath)
+        {
+            string root = Name + "\\";
+            string rootPath;
+            if (fullpath.IndexOf(root) != -1)
```
Comment (jon-r-abbott) [inline src/Projects/Model/ProjectModel.cs]: Assuming case matters?

File: src/Projects/Model/ProjectModel.cs [hunk 29]
```diff
@@ -0,0 +1,158 @@
...
+            {
+                rootPath = fullpath.Remove(0, fullpath.IndexOf(root) + root.Length);
```
Comment (jon-r-abbott) [inline src/Projects/Model/ProjectModel.cs]: Again assuming case matters?

File: src/Projects/Model/ProjectModel.cs [hunk 30]
```diff
@@ -0,0 +1,158 @@
...
+            {
+                rootPath = fullpath;
+            }
+
+            ItemModel foundItem = ParsePathSegments(rootPath.Split('\\'));
+
+            return foundItem;
+        }
+
+        private ItemModel ParsePathSegments(string[] Path)
```
Comment (jon-r-abbott) [inline src/Projects/Model/ProjectModel.cs]: Path - parameter name lowerCamelCase

**Review by jon-r-abbott — CHANGES_REQUESTED** [review]

File: src/Projects/Model/ItemModel.cs (hunk 1 above)
Comment (MichaelLiddiard) [reply src/Projects/Model/ItemModel.cs:26]: No, this is intended to be visible to other parts/clients as needed

File: src/Projects/ProjectService.cs (hunk 11 above)
Comment (MichaelLiddiard) [reply src/Projects/ProjectService.cs]: Will throw expection if cast fails

**Review by jon-r-abbott — APPROVED** [review]

**Pull request merged**